File: js/src/samples/testDataBuilder/testUserBuilder.js
import User from "./user";

const DEFAULT_NAME = "John Smith";
const DEFAULT_ROLE = "ROLE_USER";
const DEFAULT_PASSWORD = "42";


export default class TestUserBuilder {
    constructor() {
        this.name = DEFAULT_NAME;
        this.password = DEFAULT_PASSWORD;
        this.role = DEFAULT_ROLE;
        this.login = null;
    }

    static aRegularUser() {
        return this.aUser().build();
    }

    static anAdmin() {
        return this.aUser()
            .withName("Neo")
            .withLogin("neo")
            .inAdminRole()
            .build();
    }

    static aUser() {
        return new TestUserBuilder();
    }

    static but() {
        return this.aUser()
            .inRole(this.role)
            .withName(this.name)
            .withPassword(this.password)
            .withLogin(this.login)
    }

    withName(newName) {
        this.name = newName;
        return this;
    }

    withLogin(newLogin) {
        this.login = newLogin;
        return this;
    }

    withPassword(newPassword) {
        this.password = newPassword;
        return this;
    }

    withNoPassword() {
        this.password = null;
        return this;
    }

    inUserRole() {
        return this.inRole("ROLE_USER");
    }

    inAdminRole() {
        return this.inRole("ROLE_ADMIN");
    }

    inRole(newRole) {
        this.role = newRole;
        return this;
    }

    build() {
        return new User(this.name, this.login, this.password, this.role);
    }
}

File: cs/Samples/TestDataBuilder/TestUserBuilder.cs
﻿namespace Samples.TestDataBuilder
{
	public class TestUserBuilder
	{
		public const string DEFAULT_NAME = "John Smith";
		public const string DEFAULT_ROLE = "ROLE_USER";
		public const string DEFAULT_PASSWORD = "42";
		private string name = DEFAULT_NAME;
		private string password = DEFAULT_PASSWORD;
		private string role = DEFAULT_ROLE;
		private string login;

		private TestUserBuilder()
		{
		}

		public static TestUserBuilder AUser()
		{
			return new TestUserBuilder();
		}

		public TestUserBuilder WithName(string newName)
		{
			name = newName;
			return this;
		}

		public TestUserBuilder WithLogin(string newLogin)
		{
			login = newLogin;
			return this;
		}

		public TestUserBuilder WithPassword(string newPassword)
		{
			password = newPassword;
			return this;
		}

		public TestUserBuilder WithNoPassword()
		{
			password = null;
			return this;
		}

		public TestUserBuilder InUserRole()
		{
			return InRole("ROLE_USER");
		}

		public TestUserBuilder InAdminRole()
		{
			return InRole("ROLE_ADMIN");
		}

		public TestUserBuilder InRole(string newRole)
		{
			role = newRole;
			return this;
		}

		public TestUserBuilder But()
		{
			return AUser()
				.InRole(role)
				.WithName(name)
				.WithPassword(password)
				.WithLogin(login);
		}

		public User Build()
		{
			return new User(name, login, password, role);
		}

		public static User ARegularUser()
		{
			return AUser().Build();
		}

		public static User AnAdmin()
		{
			return AUser()
				.WithName("Neo")
				.WithLogin("neo")
				.InAdminRole()
				.Build();
		}
	}
}